### PR TITLE
runner: Sort history items by relevance

### PR DESCRIPTION
**Pull request by palinek**
Also add posibility to configure position of history items (first/last)

### CODE REVIEW

**Comment by luis-pereira** [comment]
GTM.
@palinek What's your opinion about allowing the user to choose ?

**Comment by palinek** [comment]
> What's your opinion about allowing the user to choose ?

To choose what?

**Comment by luis-pereira** [comment]
@palinek Never mind. I was talking about the Ui stuff, that I only picked up on it now.

File: commanditemmodel.cpp [hunk 1]
```diff
@@ -142,8 +143,25 @@ bool CommandItemModel::lessThan(const QModelIndex &left, const QModelIndex &righ
 
     if (mOnlyHistory)
         return left.row() < right.row();
-    else
-        return QSortFilterProxyModel::lessThan(left, right);
+
+    HistoryItem const * i_left = dynamic_cast<HistoryItem const *>(mSourceModel->command(left));
```
Comment (paulolieuthier) [inline commanditemmodel.cpp]: Why not using `static_cast` here?

Comment (palinek) [inline commanditemmodel.cpp]: cite:
```
2) If new_type is a pointer or reference to some class D and the type of expression is a pointer or reference to its non-virtual base B, static_cast performs a downcast. Such static_cast makes no runtime checks to ensure that the object's runtime type is actually D, and may only be used safely if this precondition is guaranteed by other means, such as when implementing static polymorphism. Safe downcast may be done with dynamic_cast.
```

But we realy need the runtime check here

Comment (paulolieuthier) [inline commanditemmodel.cpp]: I see now, `mSourceModel->command()` may return a command item of another type.

**Comment by paulolieuthier** [comment]
Just tested. Works very well. GTM.

**Comment by pmattern** [comment]
I can't seem to get this working here.
Findings: Launching GUI `lxqt-runner` and filing "lxqt-config" yields a list of binaries `lxqt-config*` represented by entries from their desktop entry files. But the order of items in this list stays exactly the same after launching say the third item by clicking, quiting the corresponding application and starting over by launching binary `lxqt-runner` several times. It does not matter whether or not "Show history first" in LXQt runner's configuration dialogue is checked.

**Comment by palinek** [comment]
This PR wasn't about prioritizing of most used applications (by the way I hate such behaviour). This is about sorting the history by relevance of match (heuristic).
Note that in history are saved only "custom commands", not applications provided by *.desktop.

The heuristic for sorting does assume that the further the match begins the more relevat the match is.
E.g.:
- in history are items `sqldeveloper`, `juffed ~/tmp/sql/example.sql`
- you type `sql` into runner
- the heuristic estimates that the `sqldeveloper` item is more relevant than the `juffed ~/tmp/sql/example.sql`

**Comment by pmattern** [comment]
Confirmed. I was assuming the link set by @jubalh in lxde/lxqt#719 pointing to this PR should indicate it's a solutions to this issue which it isn't.

The fact that the history is only covering invocations which are not corresponding to desktop entry files was totally new to me.
Is it already documented somewhere?

**Comment by palinek** [comment]
> Is it already documented somewhere?

I'm not aware of. Based on the code it is the only intention of the history is to store custom commands. There is no need to store regular \*.desktop provided apps, because those appear in the list on their own.

**Pull request merged**